File: Day1/solp2.py
def calculate_similarity_score(left_list, right_list):
    # Convert right list to a frequency dictionary for O(1) lookups
    right_freq = {}
    for num in right_list:
        right_freq[num] = right_freq.get(num, 0) + 1
    
    # Calculate similarity score
    total_score = 0
    for num in left_list:
        # Multiply each left number by its frequency in right list
        frequency = right_freq.get(num, 0)
        total_score += num * frequency
    
    return total_score

# Parse the input data
left_numbers = []
right_numbers = []

# Read from inputs.txt
with open('inputs.txt', 'r') as file:
    for line in file:
        if line.strip():
            left, right = map(int, line.split())
            left_numbers.append(left)
            right_numbers.append(right)

result = calculate_similarity_score(left_numbers, right_numbers)
print(f"Similarity score: {result}")

File: Day1/solp2.swift
import Foundation

func calculateSimilarityScore(leftList: [Int], rightList: [Int]) -> Int {
    // Convert right list to a frequency dictionary for O(1) lookups
    var rightFreq: [Int: Int] = [:]
    for num in rightList {
        rightFreq[num, default: 0] += 1
    }

    // Calculate similarity score
    var totalScore = 0
    for num in leftList {
        // Multiply each left number by its frequency in right list
        let frequency = rightFreq[num] ?? 0
        totalScore += num * frequency
    }

    return totalScore
}

// Parse the input data
var leftNumbers: [Int] = []
var rightNumbers: [Int] = []

// Read from inputs.txt
if let contents = try? String(contentsOfFile: "inputs.txt", encoding: .utf8) {
    let lines = contents.components(separatedBy: .newlines)
    for line in lines {
        if !line.isEmpty {
            let numbers = line.split(separator: " ").compactMap { Int($0) }
            if numbers.count == 2 {
                leftNumbers.append(numbers[0])
                rightNumbers.append(numbers[1])
            }
        }
    }
}

let result = calculateSimilarityScore(leftList: leftNumbers, rightList: rightNumbers)
print("Similarity score: \(result)")
